http-client/Network/HTTP/Client/Connection.hs

```diff
@@ -184,10 +187,58 @@ openSocket tweakSocket addr =
             NS.connect sock (NS.addrAddress addr)
             return sock)
 
+-- Pick up an IP using an approximation of the happy-eyeballs algorithm:
+-- https://datatracker.ietf.org/doc/html/rfc8305
+-- 
 firstSuccessful :: [NS.AddrInfo] -> (NS.AddrInfo -> IO a) -> IO a
-firstSuccessful []     _  = error "getAddrInfo returned empty list"
-firstSuccessful (a:as) cb =
-    cb a `E.catch` \(e :: E.IOException) ->
-        case as of
-            [] -> E.throwIO e
-            _  -> firstSuccessful as cb
+firstSuccessful []        _  = error "getAddrInfo returned empty list"
+firstSuccessful addresses cb = do     
+    results <- newTVarIO []
+    let totalSize = length addresses
+    withAsync (tryAddresses addresses results) $ \_ -> 
+        waitForResult results totalSize   
+  where    
+
+    tryConnecting addr results = do         
+        r <- E.try $! cb addr
+        atomically $ modifyTVar' results (r:)
+     
+    -- Try to connect to every address concurrently with a delay.
+    -- Whichever succeeds first wins.
+    tryAddresses adresses results = go adresses []
+      where
+        go [addr] asyncs = do 
+            tryConnecting addr results
+            waitForThem asyncs     
+
+        go (addr : addrs) asyncs = 
+            withAsync (tryConnecting addr results) $ \a -> do             
+                -- connection attempt delay 
+                -- https://datatracker.ietf.org/doc/html/rfc8305#section-5                
+                threadDelay $ 250 * 10000                
```

Comment:
I think the intend is 250 milliseconds, but that would be `250 * 1000`, not `250 * 10000`.

Reply:
Indeed! Thanks, will fix it.